### PR TITLE
RN number control in G4MT

### PR DESCRIPTION
**Pull request by goodenou**
RNs seeds in G4MT are now handled by us. We use Crypto++ to create hashes for seeds based on art::event ID (Run#, Subrun#, Event#).

### CODE REVIEW

**Comment by gaponenko** [comment]
@FNALbuild run build tests

**Comment by gaponenko** [comment]
@FNALbuild run code checks

File: Mu2eG4/inc/Mu2eG4WorkerRunManager.hh [hunk 1]
```diff
@@ -51,7 +51,7 @@ namespace mu2e {
     void initializeUserActions(const G4ThreeVector& origin_in_world);
     void initializeRun(art::Event* art_event);
     void processEvent(art::Event*);
-    G4Event* generateEvt(G4int i_event);
+    G4Event* generateEvt(std::string i_event, std::string i_subrun, std::string i_run);
```
Comment (gaponenko) [inline Mu2eG4/inc/Mu2eG4WorkerRunManager.hh:54]: The signature of generateEvt() strikes me as non-conventional.  If we pass event,run,subrun, I expect three integer arguments.  Or a const art::EventID reference.   Alternatively, if the function only needs a unique string, pass it a string.  But I see in the implementation that i_events is converted from string back into integer.  So the single string option is not what we want, but please consider using one of the first two options I suggested.

File: Mu2eG4/src/Mu2eG4WorkerRunManager.cc [hunk 2]
```diff
@@ -11,6 +11,7 @@
 
 //Framework includes
 #include "cetlib_except/exception.h"
+#include "cstddef"
```
Comment (gaponenko) [inline Mu2eG4/src/Mu2eG4WorkerRunManager.cc:14]: suggest angular brackets on cstddef include

File: Mu2eG4/src/Mu2eG4WorkerRunManager.cc [hunk 3]
```diff
@@ -40,37 +41,54 @@
...
+  
+  
+  string CryptoPP_Hash(const string msg){
+    
+    CryptoPP::SHA3_224 hash;
+    string digest, str_hex;
+    
+    //digest is 56 hex numbers, or 28 pairs of hex numbers = 28 bytes
+    //compute the digest and keep 8 bytes of it
+    CryptoPP::StringSource(msg, true, new CryptoPP::HashFilter(hash, new CryptoPP::StringSink(digest), false, 8));
```
Comment (gaponenko) [inline Mu2eG4/src/Mu2eG4WorkerRunManager.cc:70]: Does the cryptopp library delete all the objects that are newed here?

File: Mu2eG4/src/Mu2eG4WorkerRunManager.cc [hunk 4]
```diff
@@ -313,25 +318,32 @@ namespace mu2e {
     
     if(eventHasToBeSeeded)
     {
-      long seeds[3] = { s1, s2, 0 };
+      string msg = "r" + i_run + "s" + i_subrun + "e" + i_event + salt_;
+
+      string hash_out = CryptoPP_Hash(msg);
+      vector<string> randnumstrings = {hash_out.substr(0,8), hash_out.substr(8,8)};
```
Comment (gaponenko) [inline Mu2eG4/src/Mu2eG4WorkerRunManager.cc:324]: Making sure I understand...   The CryptoPP_Hash() function uses 8 bytes of the digest, but returns it as a hex number.  That is, 16 characters long.  So substr(8,8) does not go out of bounds.  Is that right?

**Review by gaponenko — CHANGES_REQUESTED** [review]
This PR implements an important functionality - ensuring reproducibility of results when G4 is run in the MT mode.    It also looks like a nice code cleanup.     It could be merged as is, but I think changing the signature of generateEvt() would be an improvement.

File: Mu2eG4/inc/Mu2eG4WorkerRunManager.hh (hunk 1 above)
Comment (goodenou) [reply Mu2eG4/inc/Mu2eG4WorkerRunManager.hh:54]: I do need the run, subrun, and event numbers as strings in the function generateEvt in order to make the input string for the hash:
string msg = "r" + i_run + "s" + i_subrun + "e" + i_event + salt_;

However, I can use the const art::EventID reference as you suggest for the argument of the function and convert that to a string inside the function.

**Comment by gaponenko** [comment]
On Wed, 16 Dec 2020, Lisa Goodenough wrote:

> @goodenou commented on this pull request.
> 
> 
> 
> > @@ -51,7 +51,7 @@ namespace mu2e {
>      void initializeUserActions(const G4ThreeVector& origin_in_world);
>      void initializeRun(art::Event* art_event);
>      void processEvent(art::Event*);
> -    G4Event* generateEvt(G4int i_event);
> +    G4Event* generateEvt(std::string i_event, std::string i_subrun, std::string i_run);
> 
> I do need the run, subrun, and event numbers as strings in the function generateEvt in order to make the input string for the hash:
> string msg = "r" + i_run + "s" + i_subrun + "e" + i_event + salt_;
> 
> However, I can use the const art::EventID reference as you suggest for the argument of the function and convert that to a string inside the function.

Yes, that's what I mean.  Make the string inside the function, use 
EventID in the interface.

Thank you!
Andrei

File: Mu2eG4/src/Mu2eG4WorkerRunManager.cc (hunk 2 above)
Comment (goodenou) [reply Mu2eG4/src/Mu2eG4WorkerRunManager.cc:14]: Honestly, I don't remember ever adding that include and I don't see why we need it now, so I am going to take it out.

**Review by kutschke — CHANGES_REQUESTED** [review]

File: Mu2eG4/src/SConscript [hunk 5]
```diff
@@ -19,7 +19,10 @@ helper=mu2e_helper(env)
 
 g4inc    = os.environ['G4INCLUDE'] + '/Geant4'
 g4libdir = os.environ['G4LIB']
-g4LibInc = [ '-L'+g4libdir, '-I'+g4inc ]
+g4LibInc = [ '-L'+g4libdir, '-I'+g4inc]
+cryptoppinc = os.environ['CRYPTOPP_INC']
+cryptopplibdir = os.environ['CRYPTOPP_LIB']
+g4LibInc = [ '-L'+g4libdir, '-I'+g4inc, '-L'+cryptopplibdir, '-I'+cryptoppinc ]
 g4_version = os.environ['GEANT4_VERSION']
```
[the remark this answers is not in the document]
Comment (kutschke) [reply Mu2eG4/src/SConscript:26]: Thanks.

**Review by goodenou — COMMENTED** [review]
I have responded to all of the suggestions and comments.

**Pull request closed without merging**